### PR TITLE
#286487 - Created Action on FlowManager to Redirect to Attendance

### PR DESCRIPTION
**Pull request by BeltrameJP**
Card: https://dev.azure.com/curupira/BLiP/_workitems/edit/286487

### CODE REVIEW

File: src/Take.Blip.Builder/FlowManager.cs [hunk 1]
```diff
@@ -154,6 +157,8 @@ public async Task ProcessInputAsync(Message message, Flow flow, CancellationToke
                         // Load the user context
                         var context = _contextProvider.CreateContext(userIdentity, ownerIdentity, lazyInput, flow);
 
+                        var stateIdContext = await context.GetVariableAsync("stateId", cancellationToken);
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:160]: Is this being used?

File: src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+            AttendanceRedirectHandler = new AttendanceRedirectHandler(HelpDeskExtension, Configuration);
+
+            FromIdentity = new Identity() { Domain = "msging.net", Name = "test" };
+            UserIdentity = new Identity() { Domain = "msging.net", Name = "user" };
+
+            CancellationToken = default;
+        }
+
+        public void Dispose()
```
Comment (caiopsouza) [inline src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs:38]: Is this being called? Do you have to implement `IDisposable`?

**Review by caiopsouza — APPROVED** [review]

File: src/Take.Blip.Builder/AttendanceRedirectHandler.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+using Take.Blip.Builder.Hosting;
+using Take.Blip.Client.Extensions.HelpDesk;
+using Takenet.Iris.Messaging.Resources;
+
+namespace Take.Blip.Builder
+{
+    public class AttendanceRedirectHandler : IAttendanceRedirectHandler
+    {
+
+        public static string ATTENDANCE_REDIRECT_KEY = "attendanceRedirect";
```
Comment (EricJangola) [inline src/Take.Blip.Builder/AttendanceRedirectHandler.cs:13]: This keys are being used in another class? Make them private string only if not.

File: src/Take.Blip.Builder/AttendanceRedirectHandler.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                return;
+
+            await context.SetVariableAsync(HELP_DESK_HAS_TICKET, true.ToString(), cancellationToken);
+
+            await CreateTicketWithRedirectAsync(attendanceRedirect, fromIdentity, userIdentity, cancellationToken);
+
+            await context.DeleteVariableAsync(ATTENDANCE_REDIRECT_KEY, cancellationToken);
+        }
+
+
```
Comment (EricJangola) [inline src/Take.Blip.Builder/AttendanceRedirectHandler.cs:44]: remove unecessary enter

File: src/Take.Blip.Builder/AttendanceRedirectHandler.cs (hunk 3 above)
Comment (BeltrameJP) [reply src/Take.Blip.Builder/AttendanceRedirectHandler.cs:13]: its beeing used on tests

File: src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs (hunk 2 above)
Comment (BeltrameJP) [reply src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs:38]: I'd to implement it for using IClassFixture. As I used Async tests, its a good practice for XUnit to use it.

File: src/Take.Blip.Builder/FlowManager.cs (hunk 1 above)
Comment (BeltrameJP) [reply src/Take.Blip.Builder/FlowManager.cs:160]: Its not. Removed.

File: src/Take.Blip.Builder/AttendanceRedirectHandler.cs (hunk 4 above)
Comment (BeltrameJP) [reply src/Take.Blip.Builder/AttendanceRedirectHandler.cs:44]: Fixed.

File: src/Samples/Builder.Console/application.json [hunk 5]
```diff
@@ -1,275 +1,275 @@
 ﻿{
-    "identifier": "enginelocaltimezone",
-    "accessKey": "Q2FFZVBlc3l1eVZ4VmplakRVbWY=",
-    "hostName": "hmg-tcp.msging.net",
-    "port": 443,
-    "messageReceivers": [
+  "identifier": "enginelocaltimezone",
```
Comment (EricJangola) [inline src/Samples/Builder.Console/application.json:2]: remove the changes here

File: src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
+﻿namespace Take.Blip.Builder.UnitTests
+{
+    public class AttendanceRedirectHandlerFixture //: IDisposable
+    {/*
+        public IContext Context { get; private set; }
```
Comment (andrebires) [inline src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs:5]: Do not commit "commented" code. Uncomment or remove it.

File: src/Take.Blip.Builder/FlowManager.cs [hunk 7]
```diff
@@ -175,6 +178,9 @@ public async Task ProcessInputAsync(Message message, Flow flow, CancellationToke
                         // Create trace instances, if required
                         var (stateTrace, stateStopwatch) = _traceManager.CreateStateTrace(inputTrace, state);
 
+                        // Redirect to Attendance Action
+                        await _attendanceRedirectHandler.RedirectToAttendanceAsync(message.From.ToIdentity(), userIdentity, context, cancellationToken);
```
Comment (andrebires) [inline src/Take.Blip.Builder/FlowManager.cs:182]: The FlowManager class should not know anything about the specific flow actions ("Attendence" in the case). Move to another scope (outside?)

**Review by andrebires — CHANGES_REQUESTED** [review]

File: src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs (hunk 6 above)
Comment (BeltrameJP) [reply src/Take.Blip.Builder.UnitTests/AttendanceRedirectHandlerTests.cs:5]: The code its not like this. It was simply done for test one of the check steps. I'll rever this comment code.

File: src/Take.Blip.Builder/AttendanceRedirectHandler.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Lime.Protocol;
+using Take.Blip.Builder.Hosting;
+using Take.Blip.Client.Extensions.HelpDesk;
+using Takenet.Iris.Messaging.Resources;
+
+namespace Take.Blip.Builder
+{
+    public class AttendanceRedirectHandler : IAttendanceRedirectHandler
```
[the remark this answers is not in the document]
Comment (BeltrameJP) [reply src/Take.Blip.Builder/AttendanceRedirectHandler.cs:10]: There's two reasons behind why Attendance was choose for this: 
- It was previously used on Desk
- The redirect action will also feature queues, so agents wont be the only option for users in a future.

Adding this context, I don't really know if the feature name still wrong. But, if its wrong, we need to think if its worth it changing it, not just in this context but also in Broadcast and Active Campaign.

**Comment by BeltrameJP** [comment]
Closed due to SDK was the wrong place to deal with the problem. Changes will be migrated to TemplateHosting application

**Pull request closed without merging**